### PR TITLE
AddSubscribeAttribute/Event API

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
Fix ReadEvent for issue #11951 and extend ReadEvent with subscribe capability, and extend ReadAttribute with subscription capability

#### Change overview
See above

#### Testing
Add unit test for subscribe

### CODE REVIEW

**Comment by yunhanw-google** [comment]
maybe we need separate function for subscribe since we need additional callback for subscription established

**Review by yufengwangca — APPROVED** [review]

File: src/controller/ReadInteraction.h [hunk 1]
```diff
@@ -45,7 +45,10 @@ template <typename DecodableAttributeType>
 CHIP_ERROR ReadAttribute(Messaging::ExchangeManager * aExchangeMgr, const SessionHandle sessionHandle, EndpointId endpointId,
                          ClusterId clusterId, AttributeId attributeId,
                          typename TypedReadAttributeCallback<DecodableAttributeType>::OnSuccessCallbackType onSuccessCb,
-                         typename TypedReadAttributeCallback<DecodableAttributeType>::OnErrorCallbackType onErrorCb)
+                         typename TypedReadAttributeCallback<DecodableAttributeType>::OnErrorCallbackType onErrorCb,
```
Comment (andy31415) [inline src/controller/ReadInteraction.h:48]: could we set some of these arguments as some 'params' object?

This  method now has 11 arguments out of which 2 are just 'integers' and one is a bool (so it is generally hard to read `Foo(true)` and figure out what true actually does, like `ReadAttribute(...., true, ...)` nobody will obviously figure out that true means keepSubscription.

Comment (bzbarsky-apple) [reply src/controller/ReadInteraction.h:48]: This method is not really the public API (the templates over the attribute info are; this is a utility helper), but more generally it doesn't make sense to put subscriptions into an API called ReadAttribute.

We should have a Subscribe public API that perhaps calls this method (which could get renamed) and takes subscription info as a params struct or something.

File: src/controller/ReadInteraction.h [hunk 2]
```diff
@@ -139,19 +162,13 @@ template <typename AttributeTypeInfo>
...
-                                                                    AttributeTypeInfo::GetAttributeId(), onSuccessCb, onErrorCb);
-}
-
-template <typename EventTypeInfo>
-CHIP_ERROR ReadEvent(Messaging::ExchangeManager * aExchangeMgr, const SessionHandle sessionHandle, EndpointId endpointId,
-                     typename TypedReadEventCallback<typename EventTypeInfo::DecodableType>::OnSuccessCallbackType onSuccessCb,
-                     typename TypedReadEventCallback<typename EventTypeInfo::DecodableType>::OnErrorCallbackType onErrorCb)
+              typename TypedReadAttributeCallback<typename AttributeTypeInfo::DecodableType>::OnErrorCallbackType onErrorCb,
+              uint16_t aMinIntervalFloorSeconds = 0, uint16_t aMaxIntervalCeilingSeconds = 0, bool aKeepSubscription = false,
+              app::ReadClient::InteractionType aType = app::ReadClient::InteractionType::Read)
```
Comment (bzbarsky-apple) [inline src/controller/ReadInteraction.h:167]: Again, these last bits don't belong in a public ReadAttribute API.

**Review by bzbarsky-apple — COMMENTED** [review]
Some other thoughts:

1. With this API, how does the client cancel or tear down the subscription?
2. It seems like we should expose an API to do wildcard subscription (at least of the "all endpoints, all clusters, all attributes" variety), since that's a pretty common use case.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
See comments above.

**Comment by yunhanw-google** [comment]
Based upon the conversation with @bzbarsky-apple , Boris would take the further work on it and corresponding application integration.

**Pull request closed without merging**